[WIP] Add fused types, attempt to use in denoise

AetherUnbound opened the pull request:
## Description
This is my first stab at adding the fused types from #3253 in order to fix the issue identified in #3449. I'm totally new to Cython as a whole, so I'm open to assistance with this one! 

## Checklist
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests


## For reviewers

(Don't remove the checklist below.)

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

AetherUnbound commented:
Looking at this now, I'm seeing a _lot_ of similarities with #3253 :man_facepalming: specifically with `interpolation.pxd`. @jni, what's the best move here?

hmaarrfk commented:
Let me cleanup that PR. you should be able to cherry-pick the commits really soon and incldue them in your PR.

hmaarrfk commented:
@AetherUnbound Feel free to just cherry pick the commits you need from my fork.

```
git remote add hmaarrfk https://github.com/hmaarrfk/scikit-image.git
git fetch hmaarrfk
git cherry-pick THE_COMMITS YOU WANT
```

I tested the interpolation fused code with the test suite so hopefully it shouldn't cause you lots of problems.

No guarantees that it will work, but you know, if it helps.

you might also want to have a look at this PR: https://github.com/scikit-image/scikit-image/pull/3254
In case the discussion there helps you.

jni commented:
@AetherUnbound good stuff! I think the best approach now is:

- wait for #3254 to get merged
- cherry pick @hmaarrfk's newly-minted commits from #3253 
- apply a new commit on top of those for the denoise file. You can probably cherry-pick your own commit using `--strategy=ours` (meaning: keep the changes from @hmaarrfk instead of your own if there is a conflict)
- push that set of commits to a new PR or `--force-with-lease` push to this branch to update this PR

Let us know if you need any git-help!

AetherUnbound commented:
Thanks for all the help/suggestions guys - I've cherry-picked the commits but now when I try to run `cython skimage/_shared/fused_numerics.pdx` I get this:
```
 aether   bugfix/float-cast  ~  git  scikit-image  cython skimage/_shared/fused_numerics.pxd 

Error compiling Cython file:
------------------------------------------------------------
...
cimport numpy as cnp
import numpy as np

ctypedef fused np_ints:
^
------------------------------------------------------------

skimage/_shared/fused_numerics.pxd:4:0: 'np_ints' redeclared 
```
I'm assuming this has to do with the md5/c compiled stuff.

AetherUnbound commented:
I've changed `double` to be `np_floats` in `_denoise_cy.pxd` (do I need to change instances of `np.double` to be something else?). The file itself compiles, but when I try and run the sample function I supplied in #3449 I get this:
```
In [1]: import numpy as np
   ...: import skimage.restoration
   ...: 
   ...: sigma = 16.0
   ...: sigma_range = .2
   ...: np.random.seed(0)
   ...: image = np.random.uniform(size=(100, 100)).astype(np.float32)
   ...: mask = np.ones(image.shape, bool)
   ...: mask[40:60, 45:65] = False
   ...: expected = skimage.restoration.denoise_bilateral(
   ...:     image=image,
   ...:     multichannel=False,
   ...:     sigma_color=sigma_range,
   ...:     sigma_spatial=sigma,
   ...: )
   ...: 
   ...: 


---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
~/git/scikit-image/skimage/__init__.py in <module>()
    120     try:
--> 121         from ._shared import geometry
    122         del geometry

ImportError: cannot import name 'geometry'

During handling of the above exception, another exception occurred:

ImportError                               Traceback (most recent call last)
<ipython-input-1-c4156936ec11> in <module>()
      1 import numpy as np
----> 2 import skimage.restoration
      3 
      4 sigma = 16.0
      5 sigma_range = .2

~/git/scikit-image/skimage/__init__.py in <module>()
    122         del geometry
    123     except ImportError as e:
--> 124         _raise_build_error(e)
    125 
    126     # All skimage root imports go here

~/git/scikit-image/skimage/__init__.py in _raise_build_error(e)
    102     raise ImportError("""%s
    103 It seems that scikit-image has not been built correctly.
--> 104 %s""" % (e, msg))
    105 
    106 

ImportError: cannot import name 'geometry'
It seems that scikit-image has not been built correctly.

Your install of scikit-image appears to be broken.
Try re-installing the package following the instructions at:
https://scikit-image.org/docs/stable/install.html 
```
Even after doing a `pip install -e .` (is there an easier way to just recompile the files that changed?)

hmaarrfk commented:
Can you try to remove all md5 files and compile again with python setup develop or pip install. You have to keep running those commands every change

AetherUnbound commented:
I ran the following and am still getting the same issue with the `geometry` import:
```
 aether  (e) skimage   bugfix/float-cast  ~  git  scikit-image  find . -type f -name "*.md5" | xargs rm
 aether  (e) skimage   bugfix/float-cast  ~  git  scikit-image  pip install -e .
``

hmaarrfk commented:
FYI, that MD5 PR just got pulled in. If you want you can pull upstream mster, and rebase your branch.

hmaarrfk commented on skimage/restoration/_denoise_cy.pyx:
```diff
@@ -8,17 +8,18 @@ import numpy as np
 
     cdef:
-        double[:] color_lut = np.empty(bins, dtype=np.double)
+        np_floats[:] color_lut = np.empty(bins, dtype=np.double)
```
Looking at this at a reasonable type, I can see why this is done in cython, these seem like implementation details of the algorithm. I think you might just need to change `np.double` to `np_floats`. For each version of the function, Cython will pick the same float for all instances of `np_floats` in a function.

AetherUnbound commented:
Awesome, I've rebased and changed `np.doubles` to `np_floats`, but now I get the following error:
```
    Error compiling Cython file:
    ------------------------------------------------------------
    ...
    
    
    cdef np_floats[:] _compute_color_lut(Py_ssize_t bins, np_floats sigma, np_floats max_value):
    
        cdef:
            np_floats[:] color_lut = np.empty(bins, dtype=np_floats)
                                                         ^
    ------------------------------------------------------------
    
    skimage/restoration/_denoise_cy.pyx:22:54: 'np_floats' is not a constant, variable or function identifier
    
    Error compiling Cython file:
    ------------------------------------------------------------
    ...
    
    
    cdef np_floats[:] _compute_color_lut(Py_ssize_t bins, np_floats sigma, np_floats max_value):
    
        cdef:
            np_floats[:] color_lut = np.empty(bins, dtype=np_floats)
                                                         ^
    ------------------------------------------------------------
    
    skimage/restoration/_denoise_cy.pyx:22:54: Type is not specialized
```

AetherUnbound commented:
Thanks @hmaarrfk! I'm not sure how I'd refactor denoise to pull out all of the `np.empty` or `np.zeros` calls.

I've given it a stab in cython:
```

cdef inline np_floats[:] _np_empty(Py_ssize_t shape, np_floats dtype):
    if dtype is np.float32:
        return np.empty(shape, np.float32)
    else:
        return np.empty(shape, np.float64)


cdef inline np_floats[:, :, ::1] _np_zeros(Py_ssize_t shape, np_floats dtype):
    if dtype is np.float32:
        return np.zeros(shape, np.float32)
    else:
        return np.zeros(shape, np.float64)


cdef np_floats[:] _compute_color_lut(Py_ssize_t bins, np_floats sigma, np_floats max_value):

    cdef:
        np_floats[:] color_lut = _np_empty(bins, np_floats)
        Py_ssize_t b

    sigma *= sigma
    max_value /= bins

    for b in range(bins):
        color_lut[b] = _gaussian_weight(sigma, b * max_value)

    return color_lut
```
But I'm getting the following errors:
```
Error compiling Cython file:
------------------------------------------------------------
...


cdef np_floats[:] _compute_color_lut(Py_ssize_t bins, np_floats sigma, np_floats max_value):

    cdef:
        np_floats[:] color_lut = _np_empty(bins, np_floats[:])
                                                ^
------------------------------------------------------------

skimage/restoration/_denoise_cy.pyx:36:49: 'np_floats' is not a constant, variable or function identifier

Error compiling Cython file:
------------------------------------------------------------
...


cdef np_floats[:] _compute_color_lut(Py_ssize_t bins, np_floats sigma, np_floats max_value):

    cdef:
        np_floats[:] color_lut = _np_empty(bins, np_floats[:])
                                                ^
------------------------------------------------------------

skimage/restoration/_denoise_cy.pyx:36:49: Type is not specialized

Error compiling Cython file:
------------------------------------------------------------
...


cdef np_floats[:] _compute_color_lut(Py_ssize_t bins, np_floats sigma, np_floats max_value):

    cdef:
        np_floats[:] color_lut = _np_empty(bins, np_floats[:])
                                         ^
------------------------------------------------------------

skimage/restoration/_denoise_cy.pyx:36:42: no suitable method found
```
(Sorry I'm not of more help here :sob:)

hmaarrfk commented:
Dtype is a real hack. You should basically pass sigma in.

AetherUnbound commented:
Got it, I'm passing in `sigma` now to `_np_empty` and that's working fine. However, the `_np_zeros` function is throwing `no suitable method found` for every call that's being made. Here's the function:
```
cdef inline np_floats[:, :, ::1] _np_zeros(Py_ssize_t shape, np_floats dtype):
    if dtype is np.float32:
        return np.zeros(shape, np.float32)
    else:
        return np.zeros(shape, np.float64)
```
And an example call:
```
def _denoise_bilateral(image, Py_ssize_t win_size, sigma_color,
                      np_floats sigma_spatial, Py_ssize_t bins,
                      mode, np_floats cval):
    cdef:
        np_floats min_value, max_value

    min_value = image.min()
    max_value = image.max()

    if min_value == max_value:
        return image

    # if image.max() is 0, then dist_scale can have an unverified value
    # and color_lut[<int>(dist * dist_scale)] may cause a segmentation fault
    # so we verify we have a positive image and that the max is not 0.0.
    if min_value < 0.0:
        raise ValueError("Image must contain only positive values")

    if max_value == 0.0:
        raise ValueError("The maximum value found in the image was 0.")

    image = np.atleast_3d(img_as_float(image))

    cdef:
        Py_ssize_t rows = image.shape[0]
        Py_ssize_t cols = image.shape[1]
        Py_ssize_t dims = image.shape[2]
        Py_ssize_t window_ext = (win_size - 1) / 2
        Py_ssize_t max_color_lut_bin = bins - 1

        np_floats[:, :, ::1] cimage
        np_floats[:, :, ::1] out

        np_floats[:] color_lut
        np_floats[:] range_lut

        Py_ssize_t r, c, d, wr, wc, kr, kc, rr, cc, pixel_addr, color_lut_bin
        np_floats value, weight, dist, total_weight, csigma_color, color_weight, \
               range_weight, t
        np_floats dist_scale
        np_floats[:] values
        np_floats[:] centres
        np_floats[:] total_values

    if sigma_color is None:
        csigma_color = image.std()
    else:
        csigma_color = sigma_color

    if mode not in ('constant', 'wrap', 'symmetric', 'reflect', 'edge'):
        raise ValueError("Invalid mode specified.  Please use `constant`, "
                         "`edge`, `wrap`, `symmetric` or `reflect`.")
    cdef char cmode = ord(mode[0].upper())

    cimage = np.ascontiguousarray(image)

    out = _np_zeros((rows, cols, dims), sigma_spatial)

```
And the compiler error:
```
Error compiling Cython file:
------------------------------------------------------------
...
                         "`edge`, `wrap`, `symmetric` or `reflect`.")
    cdef char cmode = ord(mode[0].upper())

    cimage = np.ascontiguousarray(image)

    out = _np_zeros((rows, cols, dims), sigma_spatial)
                  ^
------------------------------------------------------------

skimage/restoration/_denoise_cy.pyx:131:19: no suitable method found
```

AetherUnbound commented:
Got it - I totally missed that `_np_zeros` is returning a 3D array and should thus receive its shape as a tuple and not a `Py_ssize_t`. Working on that now.

AetherUnbound commented:
Turns out I wasn't close at all (cause I was checking if the parameter passed in was `np.float32`, not if it's type was `np.float32`). Here's where I'm at now: 
```
cdef inline np_floats[:] _np_empty(Py_ssize_t size, np_floats dtype):
    if np_floats is np.float32:
        return np.empty(size, np.float32)
    else:
        return np.empty(size, np.float64)


cdef inline np_floats[:, :, ::1] _np_zeros(tuple shape, np_floats[:, :, ::1] dtype):
    if np_floats is np.float32:
        return np.zeros(shape, np.float32)
    else:
        return np.zeros(shape, np.double)
```

jni commented:
I'm going to go ahead and second @hmaarrfk's original sentiment that you should aim to make all allocations in Python and then pass both input and output arrays to Cython. That will greatly simplify the code, as maintaining input dtypes in Python is super easy: `lut = np.empty(shape, dtype=image.dtype)`; `out=np.zeros_like(image)`.

jni commented:
And of course there is virtually zero performance benefit in doing this in Cython.

AetherUnbound commented:
I like that idea! The trouble is that there's all these other arrays that should also be zero (https://github.com/scikit-image/scikit-image/blob/master/skimage/restoration/_denoise_cy.pyx#L116-L122), but this is all done with shape calculations based on the `atleast_3d`'d array: https://github.com/scikit-image/scikit-image/blob/master/skimage/restoration/_denoise_cy.pyx#L81. Since most of that block is just allocating arrays and calculating shapes, should it all be moved into python?

jni commented:
@AetherUnbound yes, feel free to “refactor with abandon”, as a friend once put it. That’s what the tests are for! =) (And of course, at the end of it, please add a test with a float32 image as input!)

pep8speaks commented:
Hello @AetherUnbound! Thanks for updating the PR.

 - In the file [`skimage/restoration/_denoise.py`](https://github.com/AetherUnbound/scikit-image/blob/dd234eb2280c53ae01dbe5d3247554e37d00ec18/skimage/restoration/_denoise.py), following                are the PEP8 issues :
> [Line 5:80](https://github.com/AetherUnbound/scikit-image/blob/dd234eb2280c53ae01dbe5d3247554e37d00ec18/skimage/restoration/_denoise.py#L5): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 79 characters)
> [Line 199:80](https://github.com/AetherUnbound/scikit-image/blob/dd234eb2280c53ae01dbe5d3247554e37d00ec18/skimage/restoration/_denoise.py#L199): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)

AetherUnbound commented:
Alright! I've tried to pull all of the array allocation into python. I can't test it locally (although it does compile) because I get 
```
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
~/git/scikit-image/skimage/__init__.py in <module>()
    120     try:
--> 121         from ._shared import geometry
    122         del geometry

ImportError: cannot import name 'geometry'

During handling of the above exception, another exception occurred:

ImportError                               Traceback (most recent call last)
<ipython-input-1-c4156936ec11> in <module>()
      1 import numpy as np
----> 2 import skimage.restoration
      3 
      4 sigma = 16.0
      5 sigma_range = .2

~/git/scikit-image/skimage/__init__.py in <module>()
    122         del geometry
    123     except ImportError as e:
--> 124         _raise_build_error(e)
    125 
    126     # All skimage root imports go here

~/git/scikit-image/skimage/__init__.py in _raise_build_error(e)
    102     raise ImportError("""%s
    103 It seems that scikit-image has not been built correctly.
--> 104 %s""" % (e, msg))
    105 
    106 

ImportError: cannot import name 'geometry'
It seems that scikit-image has not been built correctly.
```
I also can't run the unit test because:
```
ImportError while importing test module '/home/aether/git/scikit-image/skimage/restoration/tests/test_denoise.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
skimage/__init__.py:135: in <module>
    from .data import data_dir
skimage/data/__init__.py:16: in <module>
    from ._binary_blobs import binary_blobs
skimage/data/_binary_blobs.py:2: in <module>
    from ..filters import gaussian
skimage/filters/__init__.py:3: in <module>
    from .edges import (sobel, sobel_h, sobel_v,
skimage/filters/edges.py:17: in <module>
    from ..restoration.uft import laplacian
skimage/restoration/__init__.py:7: in <module>
    from ._denoise import (denoise_tv_chambolle, denoise_tv_bregman,
skimage/restoration/_denoise.py:5: in <module>
    from ..restoration._denoise_cy import _denoise_bilateral, _denoise_tv_bregman, \
E   ImportError: cannot import name '_gaussian_weight' from 'skimage.restoration._denoise_cy' (/home/aether/git/scikit-image/skimage/restoration/_denoise_cy.cpython-37m-x86_64-linux-gnu.so)
```
I'm not sure if it's something that's broken or if I'm not compiling it right.

jni commented:
@AetherUnbound when I get compile errors I try to run `make clean; pip install --no-deps -U -e .`. Often it's some weird caching issue, which iirc @hmaarrfk has done a lot to address recently but I am barely following that conversation. But my crusty old-fashioned approach has never let me down, it's just slow. =)

hmaarrfk commented:
@jni, numpy won't even give you a clean command. They just say:
```
`setup.py clean` is not supported, use one of the following instead:

  - `git clean -xdf` (cleans all files)
  - `git clean -Xdf` (cleans all versioned files, doesn't touch
                      files that aren't checked into the git repo)

Add `--force` to your command to use it anyway if you must (unsupported).

```

AetherUnbound commented:
I'm not how to resolve the `no matching signature` errors, any suggestions?

hmaarrfk commented:
@AetherUnbound I just tried to compile 6685e2bb4f96a87bcff08b5a9d225bbf8c667d7d and it seemed to compile. Did you make any changes?

AetherUnbound commented:
Sorry should have been more specific - I'm getting 
```
_________________________________________________________________________________________________________ test_denoise_bilateral_nan _________________________________________________________________________________________________________

    def test_denoise_bilateral_nan():
        img = np.full((50, 50), np.NaN)
        # This is in fact an optional warning for our test suite.
        # Python 3.5 will not trigger a warning.
        with expected_warnings(['invalid|\A\Z']):
>           out = restoration.denoise_bilateral(img, multichannel=False)

skimage/restoration/tests/test_denoise.py:238: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
skimage/restoration/_denoise.py:156: in denoise_bilateral
    out, empty_dims)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   cimage = np.ascontiguousarray(image, dtype=image.dtype)
E   ValueError: Buffer dtype mismatch, expected 'float32_t' but got 'double'
``` 
when I run the unit tests, which seems to be the original problem :disappointed:

hmaarrfk commented:
Right, ok. so start by moving even more of the numpy calls out of cython.

```
-        np_floats[:, :, ::1] cimage = np.ascontiguousarray(image)
```
should just get out of there. Do that in python, and then your cython function definition becomes

```
-def _denoise_tv_bregman(image, np_floats weight, int max_iter, np_floats eps,
-                       char isotropic, np_floats[:, :, ::1] out):
+def _denoise_tv_bregman(np_floats[:, :, ::1] image, np_floats weight,
+                        int max_iter, np_floats eps,
+                        char isotropic, np_floats[:, :, ::1] out):
```

Make sure you catch the last cimage in the function
```
-                        ) + weight * cimage[r - 1, c - 1, k]
+                        ) + weight * image[r - 1, c - 1, k]
```

hmaarrfk commented:
With those fixes, you might even be at full tests passing. not sure. I just looked through a few more things, you might want to incorporate more explicit cython definitions for _denoise

<details> <summary> Proposed patch </summary>

```patch
diff --git a/skimage/restoration/_denoise_cy.pyx b/skimage/restoration/_denoise_cy.pyx
index beba36a4..3da763cd 100644
--- a/skimage/restoration/_denoise_cy.pyx
+++ b/skimage/restoration/_denoise_cy.pyx
@@ -15,7 +15,8 @@ def _gaussian_weight(np_floats sigma_sqr, np_floats value):
     return exp(-0.5 * value * value / sigma_sqr)
 
 
-cdef np_floats[:] _compute_range_lut(Py_ssize_t win_size, np_floats sigma, np_floats[:] range_lut):
+cdef np_floats[:] _compute_range_lut(Py_ssize_t win_size, np_floats sigma,
+                                     np_floats[::1] range_lut):
 
     cdef:
         Py_ssize_t kr, kc, dr, dc
@@ -45,7 +46,9 @@ def _denoise_bilateral(np_floats[:, :, ::1] image, np_floats max_value,
                        Py_ssize_t win_size,
                        np_floats sigma_color,
                        np_floats sigma_spatial, Py_ssize_t bins, mode,
-                       np_floats cval, color_lut, range_lut, out, empty_dims):
+                       np_floats cval,
+                       np_floats[::1] color_lut, np_floats[::1] range_lut,
+                       np_floats[:, :, ::1] out, np_floats[::1] empty_dims):
     cdef:
         Py_ssize_t rows = image.shape[0]
         Py_ssize_t cols = image.shape[1]
```
</details>

Tests are passing with this for sure.

AetherUnbound commented:
@hmaarrfk You're a life saver. I could swear that your PR on my branch was what I was doing, but I must have missed something. Do you typically have to recompile everything when you make a change or just run `cython <edited-file>`?
I'll also add those changes

hmaarrfk commented:
i just rerun `python setup.py develop`. That PR we suggested rebasing on helps alot.

```
pytest --lf skimage
```
is pretty useful too.

AetherUnbound commented:
Let me add another unit test for `float32` types and I think this might be good to go! Thanks so much for the help @hmaarrfk and @jni

AetherUnbound commented:
Although just so I know - what's the difference between `np_floats[:]` and `np_floats[::1]`?

hmaarrfk commented:
[::1] means it contiguous with strides 1. https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html#c-and-fortran-contiguous-memoryviews

AetherUnbound commented:
Haha I'd hate to take credit for it :stuck_out_tongue: I'm fine leaving it in

jni commented:
:tada: this is all super-awesome! However, we have an ominous looking error in AppVeyor:

```
python35-x64/lib/site-packages/skimage/restoration/tests/test_denoise.py::test_denoise_tv_bregman_2d Command exited with code -1073741819
```

https://ci.appveyor.com/project/scikit-image/scikit-image/builds/19505999/job/549juoaxo0lgn1mt#L8039

jni commented:
Or you can ignore me if that last commit fixes it. =P

AetherUnbound commented:
@jni it likely doesn't and I don't know why that error is occurring :<

jni commented:
Ouch. This seems to suggest that we need to acquire the GIL for some of the code:

https://stackoverflow.com/questions/38973604/process-finished-with-exit-code-1073741819-0xc0000005-cython-teamspeak3

Just to double check: tests run fine on your machine?

jni commented:
I take it back: the comments under the selected answer suggest that the answer made things worse! Why is the answer ticked??? :joy:

AetherUnbound commented:
Yea, they do work on my machine!
```
 aether  (e) skimage   bugfix/float-cast  ~  git  scikit-image  pytest skimage/restoration/tests/test_denoise.py 
============================================================================================================ test session starts =============================================================================================================
platform linux -- Python 3.7.0, pytest-3.8.2, py-1.7.0, pluggy-0.7.1
rootdir: /home/aether/git/scikit-image/skimage/restoration, inifile:
plugins: cov-2.6.0
collected 39 items                                                                                                                                                                                                                           

skimage/restoration/tests/test_denoise.py .......................................                                                                                                                                                      [100%]

============================================================================================================== warnings summary ==============================================================================================================
/home/aether/programs/anaconda3/envs/skimage/lib/python3.7/site-packages/dask/config.py:12: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
  from collections import Mapping

/home/aether/programs/anaconda3/envs/skimage/lib/python3.7/site-packages/pywt/_utils.py:6: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
  from collections import Iterable

-- Docs: https://docs.pytest.org/en/latest/warnings.html
=================================================================================================== 39 passed, 2 warnings in 5.47 seconds ====================================================================================================```

jni commented:
This other thread suggests that the error code means "memory corruption", which indeed sounds so ominous! Gah!

https://github.com/explosion/spaCy/issues/799

Anyway let me play around with it... Thank you very much for your efforts!

AetherUnbound commented:
Okay, let me know if there's anything else I can do - thanks!

hmaarrfk commented on skimage/restoration/_denoise_cy.pyx:
```diff
@@ -176,22 +133,21 @@ def _denoise_tv_bregman(image, double weight, int max_iter, double eps,
 
     shape_ext = (rows2, cols2, dims)
-    u = np.zeros(shape_ext, dtype=np.double)
+    u = out.copy()
```
you should explain what this is. Can you call np.zeros_like instead?

AetherUnbound replied:
I can definitely add a comment here

hmaarrfk commented:
This kinda seems silly:
```
return np.squeeze(np.asarray(out))
```
The `asarray` is not necessary.

hmaarrfk commented:
Nevermind, I don't like `zeros_like` anymore
https://github.com/numpy/numpy/issues/9909

AetherUnbound commented:
> ```
>     # reflect image
>     u[0, 1:-1] = image[1, :]
>     u[1:-1, 0] = image[:, 1]
>     u[-1, 1:-1] = image[-2, :]
>     u[1:-1, -1] = image[:, -2]
> ```
> 
> We can't have negative indices.

The operational stuff in this isn't my code - what would you suggest here?

hmaarrfk commented:
```
#cython: wraparound=False
```
disables negative indices in cython.


```
u[1:rows-1, 1:cols-1] = image

```
might work

AetherUnbound commented:
Okay @hmaarrfk, I'll try to address that today!

hmaarrfk commented:
I was just saying that it seems more like a memory out of bounds problem, than a OS Specific one seeing that we found a place where negative indices were used.

Sometimes, the order in which memory is allocated causes these problems to pop up.

Finally, using different OSs makes use of different memory allocators as well, which also causes these problems to surface occasionally.

We've been able to use 32bit linux builds to spot a few other problems in Cython, mostly related to using int64/int32 as pointers when we weren't supposed to.

jni commented:
@hmaarrfk :man_facepalming: 

Ok but I still see @AetherUnbound's point: as far as I can tell, the diff doesn't touch the Cython wraparound call or the negative indexing, so why was it working before. **Was it** working before??

Also, should we just create `u` in Python with `np.pad`?

hmaarrfk commented:
np pad PR is held up.

Basically, there is a trade off between high level python and straight
concatenation because it is implemented in C.

I wouldn’t use pad or block for these well posed problems.



On Mon, Oct 15, 2018 at 8:00 PM Juan Nunez-Iglesias <
notifications@github.com> wrote:

> @hmaarrfk <https://github.com/hmaarrfk> 🤦‍♂️
>
> Ok but I still see @AetherUnbound <https://github.com/AetherUnbound>'s
> point: as far as I can tell, the diff doesn't touch the Cython wraparound
> call or the negative indexing, so why was it working before. *Was it*
> working before??
>
> Also, should we just create u in Python with np.pad?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-image/scikit-image/pull/3469#issuecomment-430054892>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFfmBMALj-BIss0zsuVjA95x6FCFp_Vks5ulSGlgaJpZM4XX75i>
> .
>

hmaarrfk commented:
Might be relevant to the negative index bug: https://github.com/cython/cython/issues/2539

hmaarrfk commented:
@AetherUnbound I think you should just continue your work on this PR.

It seems that the intermittent failer was caused because the `range_lut` was never initialized.

In finding that bug, it went through the code again and a few places that could be cleaned up:
1. Your cython functions should either return a value, or mutate out. They shouldn't do both like numpy does. If they do, we need explicit tests that test both code paths.
   * You will have to call squeeze in python when this change is done.
2. In `_denoise_tv_bregman`, `u`, `cu` and `out` are all the same variable. Just remove `u` and `cu` and rename other variables that refer to `u` as referring to `out`
3. When you moved `_compute_color_lut` to python, it still looks like cython code. It should either be moved back to cython, or translated to python. A similar comment holds if you want to move the range_lut calculation to python as well.

jni commented:
Closing in favour of #3486
